logproof/src/linear_relation.rs

```diff
@@ -1404,4 +1409,256 @@ mod test {
+
+        for i in 0..num_polynomials {
+            for j in 0..poly_modulus_degree {
+                let index = i * poly_modulus_degree * coeff_modulus_size + j;
+                let coeff = rns[index];
+
+                let small_coeff = if coeff > first_coefficient / 2 {
+                    ((coeff as i128) - (first_coefficient as i128)) as i64
+                } else {
+                    coeff as i64
+                };
+
+                result[i][j] = small_coeff;
+            }
+        }
+
+        result
+    }
+
+    // `DensePolynomial` defines a function called `truncate_leading_zeros` but
+    // it is private.
+    fn strip_trailing_value<T>(mut v: Vec<T>, trim_value: T) -> Vec<T>
+    where
+        T: Eq,
+    {
+        while v.last().map_or(false, |c| *c == trim_value) {
+            v.pop();
+        }
+
+        v
+    }
+
+    fn convert_to_small_coeffs(
+        coeff_modulus: &Vec<Modulus>,
+        poly_array: PolynomialArray,
+    ) -> Vec<Vec<i64>> {
+        convert_to_smallint(coeff_modulus, poly_array)
+            .into_iter()
+            .map(|v| strip_trailing_value(v, 0))
+            .collect()
+    }
+
+    fn convert_to_polynomials_by_small_coeffs<Q>(
+        coeff_modulus: &Vec<Modulus>,
+        poly_array: PolynomialArray,
+    ) -> Vec<DensePolynomial<Q>>
+    where
+        Q: Field,
+    {
+        convert_to_small_coeffs(coeff_modulus, poly_array)
+            .into_iter()
+            .map(|v| make_poly(&v))
+            .collect::<Vec<DensePolynomial<Q>>>()
+    }
+
+    fn convert_to_polynomial<Q, const N: usize>(
+        poly_array: PolynomialArray,
+    ) -> Vec<DensePolynomial<Fp<MontBackend<Q, N>, N>>>
+    where
+        Q: MontConfig<N>,
+    {
+        let bigint_values = poly_array
+            .as_multiprecision_u64s()
+            .unwrap()
+            .chunks(N)
+            .map(|x| BigInt::<N>(x.try_into().unwrap()))
+            .collect::<Vec<BigInt<N>>>();
+
+        bigint_values
+            .chunks(poly_array.poly_modulus_degree() as usize)
+            .map(|x| DensePolynomial {
+                coeffs: x
+                    .iter()
+                    .map(|y| Fp::<MontBackend<Q, N>, N>::from(*y))
+                    .collect::<Vec<Fp<MontBackend<Q, N>, N>>>(),
+            })
+            .collect()
+    }
+
+    fn mod_poly<Q>(poly_mod_degree: usize, p: &DensePolynomial<Q>) -> DensePolynomial<Q>
+    where
+        Q: Field,
+    {
+        let mut coeffs1 = p.coeffs.clone();
+        let coeffs2 = coeffs1.split_off(poly_mod_degree);
+        let coeffs1 = strip_trailing_value(coeffs1, Q::zero());
+        let coeffs2 = strip_trailing_value(coeffs2, Q::zero());
+        let coeffs2 = coeffs2.into_iter().map(|x| Q::zero() - x).collect();
+
+        DensePolynomial { coeffs: coeffs1 } + DensePolynomial { coeffs: coeffs2 }
+    }
+
+    #[test]
+    fn full_knowledge_bfv_proof() {
+        type Fq = FqSeal128_1024;
+
+        let degree = 1024;
```

Comment:
1024 is a good litmus test for you get the right answer because SEAL doesn't use the special modulus for this lattice dimension (or 2048) anyways, so your changes have no effect.

Reply:
Precisely!